Source/WebCore/dom/Element.cpp

```diff
@@ -2735,232 +2735,313 @@ void Element::didMoveToNewDocument(Document& oldDocument, Document& newDocument)
+    // directionality of this slot.
+    if (!slotElement.isInShadowTree())
+        return computeElementContainedTextAutoDirection(slotElement);
+
+    auto* nodes = slotElement.assignedNodes();
+    if (!nodes)
+        return computeElementContainedTextAutoDirection(slotElement);
+
+    for (auto& child : *nodes) {
+        // Specs: If child is a Text node and its text direction is
+        // not null, then return it.
+        if (child->isTextNode()) {
+            if (auto direction = computeTextDirectionFromText(child->textContent(true)))
+                return direction;
         }
-    };
-    updateEffectiveTextDirectionOfShadowRoot(*this);
-    for (auto it = descendantsOfType<Element>(*this).begin(); it;) {
-        Ref element = *it;
-        if (isValidDirValue(element->attributeWithoutSynchronization(dirAttr))) {
-            it.traverseNextSkippingChildren();
-            continue;
+
+        if (auto* element = dynamicDowncast<Element>(child.get())) {
+            // Specs: If child direction is not null, then return child direction.
+            if (auto direction = element->computeAutoTextDirection())
+                return direction;
         }
-        updateEffectiveTextDirectionOfShadowRoot(element);
-        Style::PseudoClassChangeInvalidation styleInvalidation(element, CSSSelector::PseudoClass::Dir, Style::PseudoClassChangeInvalidation::AnyValue);
-        element->setUsesEffectiveTextDirection(!!direction);
-        if (direction)
-            element->setEffectiveTextDirection(effectiveDirection);
-        it.traverseNext();
     }
+
+    return std::nullopt;
 }
 
-void Element::adjustDirectionalityIfNeededAfterChildrenChanged(Element* beforeChange, ChildChange::Type changeType)
+// Specs: https://html.spec.whatwg.org/multipage/dom.html#auto-directionality
+static std::optional<TextDirection> computeTextDirectionOfFormControlElement(const HTMLTextFormControlElement& textFormControlElement)
 {
-    // FIXME: This function looks suspicious.
+    if (!textFormControlElement.dirAutoUsesValue())
+        return computeElementContainedTextAutoDirection(textFormControlElement);
 
-    if (!selfOrPrecedingNodesAffectDirAuto())
-        return;
+    // Specs: The directionality of the auto-directionality form-associated
+    // element is calculated from its value() text.
+    // Specs: If element's value is not the empty string, then return 'ltr'.
+    if (auto value = textFormControlElement.value(); !value.isEmpty())
+        return computeTextDirectionFromText(value).value_or(TextDirection::LTR);
+    return std::nullopt;
+}
 
-    RefPtr<Node> oldMarkedNode;
-    if (beforeChange)
-        oldMarkedNode = changeType == ChildChange::Type::ElementInserted ? ElementTraversal::nextSibling(*beforeChange) : beforeChange->nextSibling();
+static void updateElementAndShadowEffectiveTextDirection(Element& element, std::optional<TextDirection> direction, Element* initiator = nullptr)
```

Comment:
I'd rather call this `updateEffectiveTextDirectionOfElementAndShadowTree`.

Reply:
Done.